### PR TITLE
[DwarfDump] Add new set of line-table-related statistics to llvm-dwarfdump

### PR DESCRIPTION
**Pull request by SLTozer**
This patch adds a new set of statistics to llvm-dwarfdump that provide additional information about .debug_line regarding the number of bytes covered by the line table (and how many of those are covered by line 0 entries), and the number of entries within the table and how many of those are is_stmt, unique, or unique and non-line-0 (where "uniqueness" is based on file, line, and column only).

Collectively these give a little more insight into the state of debug line information, rather than variables (as most of the dwarfdump statistics are currently oriented towards). I've added all of the stats that were useful to some degree, but I think the most generally useful stat is "unique line entries", since it gives the most straightforward indication of regressions, i.e. when the number goes down it means that fewer source lines are reachable in the program.

I've not worked hard to make this performant, under the assumption that llvm-dwarfdump is generally running offline and does not run into serious performance constraints; if this is not the case, then there may be some suitable improvements in the uniquing and file index behaviour. The names of the statistics are straightforward and descriptive, but if there is a particular style/pattern they ought to match then I'm open to changing them as well.

### CODE REVIEW

**Review by jryans — APPROVED** [review]
Looks good to me, seems quite useful to have! 😄

**Comment by jryans** [comment]
Since this adds new stats, I suppose the schema on LNT may need to be updated... Perhaps @adrian-prantl can help with that once this merges.

File: llvm/tools/llvm-dwarfdump/Statistics.cpp [hunk 1]
```diff
@@ -882,8 +900,58 @@ bool dwarfdump::collectStatsForObjectFile(ObjectFile &Obj, DWARFContext &DICtx,
...
+              CUFileMapping[FileIdx] =
+                  std::distance(Files.begin(), ExistingFile);
+            } else {
+              CUFileMapping[FileIdx] = Files.size();
+              Files.push_back(File);
+            }
+          }
+        }
+      }
+      for (auto Seq : LineTable->Sequences) {
```
Comment (djolertrk) [inline llvm/tools/llvm-dwarfdump/Statistics.cpp:927]: `const auto&` instead?

File: llvm/tools/llvm-dwarfdump/Statistics.cpp [hunk 2]
```diff
@@ -882,8 +900,58 @@ bool dwarfdump::collectStatsForObjectFile(ObjectFile &Obj, DWARFContext &DICtx,
         CrossCUReferencesToBeResolved.push_back(
             DIELocation(CUDie.getDwarfUnit(), CrossCUReferencingDIEOffset));
     }
+    if (const auto *LineTable = DICtx.getLineTableForUnit(CU.get())) {
+      auto LastFileIdxOpt = LineTable->getLastValidFileIndex();
+      // Each CU has its own file index; in order to track unique line entries
+      // across CUs, we therefore need to map each CU file index to a global
+      // file index, which we store here.
+      DenseMap<uint64_t, uint16_t> CUFileMapping;
+      if (LastFileIdxOpt) {
```
Comment (djolertrk) [inline llvm/tools/llvm-dwarfdump/Statistics.cpp:909]: an early exit here is better

File: llvm/tools/llvm-dwarfdump/Statistics.cpp [hunk 3]
```diff
@@ -856,6 +868,12 @@ bool dwarfdump::collectStatsForObjectFile(ObjectFile &Obj, DWARFContext &DICtx,
   // abstract_origin.
   FunctionDIECUTyMap AbstractOriginFnCUs;
   CrossCUReferencingDIELocationTy CrossCUReferencesToBeResolved;
+  // Line, Col, File
```
Comment (djolertrk) [inline llvm/tools/llvm-dwarfdump/Statistics.cpp:871]: Improve the comment please

**Review by djolertrk — APPROVED** [review]
LGTM (with some minor comments).

Thanks!

File: llvm/tools/llvm-dwarfdump/Statistics.cpp (hunk 2 above)
Comment (SLTozer) [reply llvm/tools/llvm-dwarfdump/Statistics.cpp:909]: Good point - `LastFileIdxOpt` doesn't technically need to be valid for the rest of the logic, but a line table without files wouldn't have any sequences either. That being said this part isn't a loop so can't be early-exited; I can move this and the LineTable variable out of the `if` block and use them together as a precondition though.

**Comment by SLTozer** [comment]
> Since this adds new stats, I suppose the schema on LNT may need to be updated... Perhaps @adrian-prantl can help with that once this merges.

My assumption at the moment would be that this is passed through as JSON, s.t. new fields will be ignored by default - so it shouldn't cause any problems at least for the existing stats, though it would be good to get these added!

Without getting ahead of things too much (I'll have a discourse post soon going more into the topic), it looks as though the unique lines metric has been degrading in clang over time, so it'd be good to get some continuous measurement.

**Comment by jryans** [comment]
> My assumption at the moment would be that this is passed through as JSON, s.t. new fields will be ignored by default - so it shouldn't cause any problems at least for the existing stats, though it would be good to get these added!

Ah yeah, I don't think it will cause problems either, I was more just thinking about ensuring the new fields are tracked over time going forward. But yeah, IIUC, there's no rush to do anything immediately after merging this because of the JSON structure as you mentioned.

> Without getting ahead of things too much (I'll have a discourse post soon going more into the topic), it looks as though the unique lines metric has been degrading in clang over time, so it'd be good to get some continuous measurement.

Ah, good to know. Would be great to have more people watching those stats as they evolve... I might be one such person soon, we'll see. 😄

**Comment by OCHyams** [comment]
Do we need any more targeted tests for the stats (for example, exercising the cu-unique-file-id code you added)? llvm/test/tools/llvm-dwarfdump/X86/locstats.ll provides some coverage, maybe that's enough to flag up future changes causing issues. I don't have a strong opinion on this but thought it should be brought up. wdyt?

**Comment by SLTozer** [comment]
> Do we need any more targeted tests for the stats (for example, exercising the cu-unique-file-id code you added)?

Could add a second test that uses multiple compile units - not sure what the best way to handle that is, but I'm sure there'll be some existing lit examples I can base it on.

**Review by adrian-prantl — APPROVED** [review]
Since this is looking at `.debug_line` — have you tested this with both DWARF 4 and DWARF 5 inputs?

File: llvm/tools/llvm-dwarfdump/Statistics.cpp [hunk 4]
```diff
@@ -882,8 +902,58 @@ bool dwarfdump::collectStatsForObjectFile(ObjectFile &Obj, DWARFContext &DICtx,
...
+    std::optional<uint64_t> LastFileIdxOpt;
+    if (LineTable)
+      LastFileIdxOpt = LineTable->getLastValidFileIndex();
+    if (LastFileIdxOpt) {
+      // Each CU has its own file index; in order to track unique line entries
+      // across CUs, we therefore need to map each CU file index to a global
+      // file index, which we store here.
+      DenseMap<uint64_t, uint16_t> CUFileMapping;
+      std::string File;
+      for (uint64_t FileIdx = 0; FileIdx <= *LastFileIdxOpt; ++FileIdx) {
```
[the remark this answers is not in the document]
Comment (SLTozer) [reply llvm/tools/llvm-dwarfdump/Statistics.cpp:914]: They are - but the method `getFileNameByIndex` will return `std::nullopt` if you fetch index 0 for DWARFv4 and earlier, and we have the correct final index for the for-loop from `getLastValidFileIndex`, so in this case this loop still does the right thing.

**Comment by SLTozer** [comment]
> Since this is looking at `.debug_line` — have you tested this with both DWARF 4 and DWARF 5 inputs?

I've updated the locstats.ll to test for DWARF versions 2-5; only DWARF version 1 results in a failure, which happens while emitting the object file (presumably because some of the DIExpressions are unsupported).

File: llvm/tools/llvm-dwarfdump/Statistics.cpp [hunk 5]
```diff
@@ -1043,6 +1113,16 @@ bool dwarfdump::collectStatsForObjectFile(ObjectFile &Obj, DWARFContext &DICtx,
   printLocationStats(J, "#local vars", LocStats.LocalVarLocStats);
   printLocationStats(J, "#local vars - entry values",
                      LocStats.LocalVarNonEntryValLocStats);
+
+  // Print line statistics for the object file.
+  printDatum(J, "#bytes with line information", LnStats.NumBytes.Value);
+  printDatum(J, "#bytes with line-0 locations", LnStats.NumLineZeroBytes.Value);
+  printDatum(J, "#line entries", LnStats.NumEntries.Value);
+  printDatum(J, "#line entries marked is_stmt", LnStats.NumIsStmtEntries.Value);
```
Comment (OCHyams) [inline llvm/tools/llvm-dwarfdump/Statistics.cpp:1121]:
```suggestion
  printDatum(J, "#line entries (is_stmt)", LnStats.NumIsStmtEntries.Value);
```
More in keeping with lines below?

**Comment by SLTozer** [comment]
I think this is about ready to merge; only thing I couldn't do was get a proper test for tracking uniqueness across multiple CUs. I can create an IR file with multiple DICompileUnits, but that doesn't mean multiple separate compile units in the output object. I think I can't make such a test without either just committing the test object directly, or linking. Is test coverage a blocker for the patch? I don't want to commit without coverage for some functionality, but I'm not sure what the preferred approach for this is.

**Comment by OCHyams** [comment]
>I think this is about ready to merge; only thing I couldn't do was get a proper test for tracking uniqueness across multiple CUs. 

It might be possible using yaml2obj? Setting up yaml2obj tests for DWARF can be pretty fiddly though. I don't want to block this patch if others are broadly happy though

**Comment by SLTozer** [comment]
> It might be possible using yaml2obj? Setting up yaml2obj tests for DWARF can be pretty fiddly though. I don't want to block this patch if others are broadly happy though

I looked at yaml2obj, couldn't figure out a way to get it to do what I need - it seems like it doesn't support debug lines, and if it does I haven't found out how to make it do so.

**Pull request merged**